src/dune_rules/merlin/merlin.ml

```diff
@@ -214,26 +214,30 @@ module Processed = struct
       Path.Set.to_list_map hidden_src_dirs ~f:(make_directive_of_path "SH")
     in
     let flags =
-      let flags =
-        match flags with
+      (* Order matters here. The flags should be communicated to Merlin in the
+         same order that they are passed to the compiler: user flags, pp flags
+         and then opens *)
+      let open_flags =
+        match opens with
         | [] -> []
-        | flags ->
-          [ make_directive "FLG" (Sexp.List (List.map ~f:(fun s -> Sexp.Atom s) flags)) ]
+        | opens ->
+          let open_flags =
+            Ocaml_flags.open_flags opens |> List.map ~f:(fun x -> Sexp.Atom x)
+          in
+          [ make_directive "FLG" (Sexp.List open_flags) ]
       in
-      let flags =
+      let other_flags =
         match pp with
-        | None -> flags
+        | None -> open_flags
         | Some { flag; args } ->
           make_directive "FLG" (Sexp.List [ Atom (Pp_kind.to_flag flag); Atom args ])
-          :: flags
+          :: open_flags
       in
-      match opens with
-      | [] -> flags
-      | opens ->
-        make_directive
-          "FLG"
-          (Sexp.List (Ocaml_flags.open_flags opens |> List.map ~f:(fun x -> Sexp.Atom x)))
-        :: flags
+      match flags with
+      | [] -> other_flags
+      | flags ->
+        make_directive "FLG" (Sexp.List (List.map ~f:(fun s -> Sexp.Atom s) flags))
+        :: other_flags
```

Comment:
I find this code somewhat hard to read since each match defaults to the one above. Wouldn't it be easier to handle them separately and do a list concat? That way the order of how the flags are ordered is visible in one line instead of having to trace through all of the matches.

Reply:
Thanks for the review @Leonidas-from-XIV I now filter a list of options in [956c989](https://github.com/ocaml/dune/pull/11503/commits/956c9894964e9a1d856d19e28af03c9f17ed9211).